kong/resolver/access.lua

```diff
@@ -32,27 +32,93 @@ local function get_host_from_url(val)
   return parsed_url.host..(port and ":"..port or "")
 end
 
--- Retrieve the API from the Host that has been requested
-function _M.execute(conf)
-  local hosts_headers = {}
+-- Find an API from a request made to nginx. Either from one of the Host or X-Host-Override headers
+-- matching the API's `public_dns`, either from the `request_uri` matching the API's `path`.
+--
+-- To perform this, we need to query _ALL_ APIs in memory. It is the only way to compare the `request_uri`
+-- as a regex to the values set in DB. We keep APIs in the database cache for a longer time than usual.
+-- @see https://github.com/Mashape/kong/issues/15 for an improvement on this.
+--
+-- @return `err`         Any error encountered during the retrieval.
+-- @return `api`         The retrieved API, if any.
+-- @return `hosts`       The list of headers values found in Host and X-Host-Override.
+-- @return `request_uri` The URI for this request.
+local function find_api()
+  local retrieved_api
+
+  -- retrieve all APIs
+  local apis_dics, err = cache.get_or_set("APIS_BY_PUBLIC_DNS", function()
+    local apis, err = dao.apis:find_all()
+    if err then
+      return nil, err
+    end
+
+    -- build dictionnaries of public_dns:api and path:apis for efficient lookup.
+    local dns_dic, path_dic = {}, {}
+    for _, api in ipairs(apis) do
+      if api.public_dns then
+        dns_dic[api.public_dns] = api
+      end
+      if api.path then
+        path_dic[api.path] = api
+      end
+    end
+    return {dns = dns_dic, path = path_dic}
+  end, 60) -- 60 seconds cache
+
+  if err then
+    return err
+  end
+
+  -- find by Host header
+  local all_hosts = {}
   for _, header_name in ipairs({"Host", constants.HEADERS.HOST_OVERRIDE}) do
-    local host = ngx.req.get_headers()[header_name]
-    if type(host) == "string" then -- single header
-      table.insert(hosts_headers, host)
-    elseif type(host) == "table" then -- multiple headers
-      for _, v in ipairs(host) do
-        table.insert(hosts_headers, v)
+    local hosts = ngx.req.get_headers()[header_name]
+    if hosts then
+      if type(hosts) == "string" then
+        hosts = {hosts}
+      end
+      -- for all values of this header, try to find an API using the apis_by_dns dictionnary
+      for _, host in ipairs(hosts) do
+        table.insert(all_hosts, host)
+        if apis_dics.dns[host] then
+          retrieved_api = apis_dics.dns[host]
+          break
+        end
       end
     end
   end
 
-  -- Find the API
-  local api, err = resolver_util.find_api(hosts_headers)
+  -- If it was found by Host, return
+  if retrieved_api then
+    return nil, retrieved_api
+  end
+
+  -- Otherwise, we look for it by path. We have to loop over all APIs and compare the requested URI.
+  local request_uri = ngx.var.request_uri
+  for path, api in pairs(apis_dics.path) do
+    local m, err = ngx.re.match(request_uri, path)
```

Comment:
It's been awhile since I've done LUA, but it looks like path when stored in the database would be something like `/status` where the `/` would be added if not present to the beginning but the regular expression matching here doesn't check where `/status` is. 

What it looks like to me is that if I created an API with the path `/status` when it gets to the comparison point it would compile the regular expression to `/\/status/` (where `/` is my boundary character). Which means that it will match a request to `/myapi/status`, `/status`, `/something/status/somethingelse/status` etc. I'm not sure how you want to approach it, but Express in NodeJS land for example uses a few configuration parameters to create their Regular Expression. You can see those in [path-to-regexp](https://www.npmjs.com/package/path-to-regexp#usage) under the options section.